### PR TITLE
Add new Web Sockets tab

### PR DESCRIPTION
**Pull request by psiinon**
Depends on https://github.com/zaproxy/zap-extensions/pull/1926
Addresses the first checkbox on #373 ie just displaying the websocket messages
Its a rip off of the History tab, and filtering seems to work.
More testing is required ;)

### CODE REVIEW

**Comment by psiinon** [comment]
I've changed the tab badges as they werent getting updated when another tab was being shown.
Think its ready to review now :)

**Comment by psiinon** [comment]
FYI I've been testing with https://www.websocket.org/echo.html now that http://browserquest.mozilla.org/ seems to be broken :/

File: src/main/zapHomeFiles/hud/drawer.html [hunk 1]
```diff
@@ -67,6 +70,58 @@
...
+                        </th>
+                    </tr>
+                </thead>
+            </table>
+            <table id="websockets-messages" class="table table-striped table-hover table-scroll table-history drawer-messages" style="min-width: fit-content;">
+                <tbody>
+                    <tr v-for="message in filteredMessages" @click="messageSelected(message.id)" :id="'message-tr-'+message.id" class="message-tr">
+                        <td class="field time"> {{ message.time }} </td>
+                        <td class="field direction"> {{ message.direction }} </td>
+                        <td class="field opcode"> {{ message.opCode }} </td>
```
Comment (thc202) [inline src/main/zapHomeFiles/hud/drawer.html]: These and following fields need some style tweaks to be shown under the respective column.

File: src/main/zapHomeFiles/hud/tools/websockets.js [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+		}
+
+		// Directed
+		if (message.tool === NAME) {
+			switch(message.action) {
+				case "buttonMenuCicked":
+					showOptions(message.tabId);
+					break;
+
+				case "showWebSocketHttpMessageDetails":
```
Comment (thc202) [inline src/main/zapHomeFiles/hud/tools/websockets.js]: Should be `showWsMessageDetails`?

File: src/other/resources/UIMessages/UIMessages.properties [hunk 3]
```diff
@@ -102,3 +102,12 @@ spider_start = Start spidering this site?
 spider_start_scope = This site is not in scope.<br>In order to spider the site you must add it to the scope.<br>Add the site to the scope and start spidering it?
 spider_stop = The spider is currently running. Would you like to stop it?
 spider_tool = Spider
+
+websockets_enable_regex = Enable RegEx
```
Comment (thc202) [inline src/other/resources/UIMessages/UIMessages.properties]: Should reuse the existing messages?

File: src/other/resources/UIMessages/UIMessages.properties [hunk 4]
```diff
@@ -102,3 +102,12 @@ spider_start = Start spidering this site?
...
+websockets_items_filtered = {0} of {1} items hidden by filters
+websockets_message_field_time = Time
+websockets_message_field_direction = Direction
+websockets_message_field_opcode = Op Code
+websockets_message_field_bytes = Bytes
+websockets_message_field_payload = Payload
+websockets_tool = Web Sockets
```
Comment (thc202) [inline src/other/resources/UIMessages/UIMessages.properties]: Single word?

Comment (psiinon) [reply src/other/resources/UIMessages/UIMessages.properties]: yep

File: src/main/zapHomeFiles/hud/tools/websockets.js (hunk 2 above)
Comment (psiinon) [reply src/main/zapHomeFiles/hud/tools/websockets.js]: Will change display.js to as I think the longer version is better in this case

**Comment by psiinon** [comment]
Hopefully fixed all those comments.

**Comment by thc202** [comment]
Thanks!

File: src/main/zapHomeFiles/hud/tools/websockets.js [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+ */
+
+var WebSockets = (function() {
+
+	// Constants
+	var NAME = "websockets";
+	var LABEL = I18n.t("websockets_tool");
+	var ICONS = {};
+        ICONS.CLOCK = "clock.png";	// Not actually shown anywhere, should be changed if it is going to be shown
+    var tool = {};
```
Comment (dscrobonia) [inline src/main/zapHomeFiles/hud/tools/websockets.js]: lots of tab formatting issues, which I see are also in the history tool. We really need to get a linter.

File: src/main/zapHomeFiles/hud/tools/websockets.js [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+	        message.direction = event.detail.direction;
+	        message.length = event.detail.length;
+	        message.messageSummary = event.detail.messageSummary;
+	        message.opCode = event.detail.opCode + '=' + event.detail.opCodeString;
+			message.id = event.detail.messageId;
+	
+	        utils.messageAllTabs('drawer', {action: 'updateWebSockets', messages: [message]})
+	            .catch(utils.errorHandler);
+	
+			tool.messages.push(message);
```
Comment (dscrobonia) [inline src/main/zapHomeFiles/hud/tools/websockets.js]: Interesting to note here that we don't call `utils.loadTool` before calling `utils.writeTool`. (thats how it is in the history tool as well) Technically thats how all tools should work if there wasn't that FF bug. Anyways just something to point out. I think its totally fine though.

File: src/main/zapHomeFiles/hud/tools/websockets.js (hunk 5 above)
Comment (dscrobonia) [reply src/main/zapHomeFiles/hud/tools/websockets.js]: feel free to ignore this if you want. would be great to clean up the formatting tho if you can.

Comment (psiinon) [reply src/main/zapHomeFiles/hud/tools/websockets.js]: Have you definitely got the latest code? There were formatting issues before but I think they're fixed now.

**Comment by psiinon** [comment]
Arg, you're right, the HTTP messages are no longer displayed :( I'll fix that.

**Comment by psiinon** [comment]
I just needed to merge the latest changes from develop.
I've also reformatted websockets.js
@dscrobonia that looking ok now?

**Comment by psiinon** [comment]
Just added the missing changelog entry :)

**Comment by kingthorin** [comment]
Not that it should stop you by any means, but there's a pending changelod PR: https://github.com/zaproxy/zap-hud/pull/415 one way or another something will endup conflicting.

**Comment by psiinon** [comment]
Oh yeah, have no doubt it will have conflicts ;)
But I can fix those and merge if everyones now happy with it.

**Comment by psiinon** [comment]
Rebased the commits - will merge assuming the tests all pass :)

**Pull request merged**